File: insurance/app/management/service/LifeInsuranceService.java
package com.insurance.app.management.service;

import com.insurance.app.auth.entity.Users;
import com.insurance.app.auth.repository.UserRepository;
import com.insurance.app.catalog.MessageUtil;
import com.insurance.app.catalog.exception.ResourceNotFoundException;
import com.insurance.app.management.DTO.LifeInsuranceDTO;
import com.insurance.app.management.entity.LifeInsurance;
import com.insurance.app.management.repository.LifeRepository;
import com.insurance.app.purchase.entity.PolicyPurchase;
import com.insurance.app.purchase.repository.PolicyPurchaseRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class LifeInsuranceService {

    private final LifeRepository lifeRepository;
    private final PolicyPurchaseRepository purchaseRepository;
    private final UserRepository userRepository;
    private final MessageUtil messageUtil;

    public LifeInsuranceDTO addLife(LifeInsuranceDTO dto) {
        Users user = userRepository.findById(dto.getUser_id())
                .orElseThrow(() -> new ResourceNotFoundException(
                        messageUtil.getMessage("user.not.found")));

        LifeInsurance life = convertToEntity(dto);
        double premium = computePremium(dto);

        life.setPremium(premium);
        life.setStatus("PENDING");
        life.setUser(user);
        life.setPurchase(null);

        return convertToDto(lifeRepository.save(life));
    }

    public LifeInsuranceDTO confirmPurchase(Long lifeId, PolicyPurchase purchase) {
        LifeInsurance l = lifeRepository.findById(lifeId)
                .orElseThrow(() -> new ResourceNotFoundException(
                        messageUtil.getMessage("life.not.found")));
        l.setPurchase(purchase);
        l.setStatus("CONFIRMED");
        return convertToDto(lifeRepository.save(l));
    }

    public void cancelPendingLife(Long id) {
        LifeInsurance l = lifeRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException(
                        messageUtil.getMessage("life.not.found")));
        if ("PENDING".equals(l.getStatus())) {
            lifeRepository.delete(l);
        }
    }

    public LifeInsuranceDTO updateLife(Long id, LifeInsuranceDTO dto) {
        LifeInsurance l = lifeRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException(
                        messageUtil.getMessage("life.not.found")));

        l.setAge(dto.getAge());
        l.setGender(dto.getGender());
        l.setSumAssured(dto.getSumAssured());
        l.setPolicyTerm(dto.getPolicyTerm());
        l.setSmoker(dto.isSmoker());
        l.setOccupationRisk(dto.getOccupationRisk());

        double premium = computePremium(dto);
        l.setPremium(premium);

        return convertToDto(lifeRepository.save(l));
    }

    public List<LifeInsuranceDTO> getAll() {
        return lifeRepository.findAll()
                .stream()
                .map(this::convertToDto)
                .collect(Collectors.toList());
    }

    private LifeInsurance convertToEntity(LifeInsuranceDTO dto) {
        return LifeInsurance.builder()
                .age(dto.getAge())
                .gender(dto.getGender())
                .sumAssured(dto.getSumAssured())
                .policyTerm(dto.getPolicyTerm())
                .smoker(dto.isSmoker())
                .occupationRisk(dto.getOccupationRisk())
                .build();
    }

    public void deleteLife(Long id) {
        if (!lifeRepository.existsById(id)) {
            throw new ResourceNotFoundException(
                    messageUtil.getMessage("life.not.found.with.id", id));
        }
        lifeRepository.deleteById(id);
    }

    public void cancelLifePolicy(Long id) {
        LifeInsurance life = lifeRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException(
                        messageUtil.getMessage("life.not.found")));

        if ("CONFIRMED".equals(life.getStatus())) {
            life.setStatus("CANCELLED");
            lifeRepository.save(life);

            if (life.getPurchase() != null) {
                PolicyPurchase purchase = life.getPurchase();
                purchase.setStatus("CANCELLED");
                purchaseRepository.save(purchase);
            }
        } else if ("PENDING".equals(life.getStatus())) {
            lifeRepository.delete(life);
        }
    }

    private LifeInsuranceDTO convertToDto(LifeInsurance life) {
        return LifeInsuranceDTO.builder()
                .id(life.getId())
                .age(life.getAge())
                .gender(life.getGender())
                .sumAssured(life.getSumAssured())
                .policyTerm(life.getPolicyTerm())
                .smoker(life.isSmoker())
                .occupationRisk(life.getOccupationRisk())
                .premium(life.getPremium())
                .user_id(life.getUser().getId())
                .build();
    }

    public double computePremium(LifeInsuranceDTO dto) {
        int age = dto.getAge();
        if (age < 18 || age > 70) {
            throw new IllegalArgumentException(
                    messageUtil.getMessage("life.age.invalid"));
        }

        double basePremium = (dto.getSumAssured() / 1000) * 0.5;
        basePremium *= (dto.getPolicyTerm() / 10.0);

        double ageMultiplier = 1.0;
        if (age <= 25) ageMultiplier = 1.0;
        else if (age <= 35) ageMultiplier = 1.1;
        else if (age <= 45) ageMultiplier = 1.2;
        else if (age <= 55) ageMultiplier = 1.4;
        else if (age <= 65) ageMultiplier = 1.6;
        else ageMultiplier = 2.0;

        basePremium *= ageMultiplier;

        if (dto.isSmoker()) basePremium *= 1.3;

        if ("high".equalsIgnoreCase(dto.getOccupationRisk())) {
            basePremium *= 1.4;
        } else if ("medium".equalsIgnoreCase(dto.getOccupationRisk())) {
            basePremium *= 1.2;
        }

        return roundToTwoDecimal(basePremium);
    }

    private double roundToTwoDecimal(double value) {
        return Math.round(value * 100.0) / 100.0;
    }
}

File: src/app/services/life-insurance.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import { AuthService } from './auth.service';

export interface LifeInsuranceDTO {
  id?: number;
  age: number;
  gender: string;
  sumAssured: number;
  policyTerm: number;
  smoker: boolean;
  occupationRisk: string;
  premium?: number;
  user_id: number;
  status?: string;
}

export interface LifeInsurance {
  id: number;
  age: number;
  gender: string;
  sumAssured: number;
  policyTerm: number;
  smoker: boolean;
  occupationRisk: string;
  premium: number;
  status: string;
  user: any;
  purchase?: any;
}

@Injectable({
  providedIn: 'root'
})
export class LifeInsuranceService {
  private apiUrl = `${environment.apiUrl}/api/customer/life`;

  constructor(
    private http: HttpClient,
    private authService: AuthService
  ) {}

  addLife(dto: LifeInsuranceDTO): Observable<LifeInsuranceDTO> {
    return this.http.post<LifeInsuranceDTO>(`${this.apiUrl}/addLife`, dto, {
      headers: this.authService.getAuthHeaders()
    });
  }

  confirmPurchase(lifeId: number, request: any): Observable<LifeInsuranceDTO> {
    return this.http.post<LifeInsuranceDTO>(`${this.apiUrl}/confirm/${lifeId}`, request, {
      headers: this.authService.getAuthHeaders()
    });
  }

  cancelPendingLife(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/cancel/${id}`, {
      headers: this.authService.getAuthHeaders()
    });
  }

  updateLife(id: number, dto: LifeInsuranceDTO): Observable<LifeInsuranceDTO> {
    return this.http.put<LifeInsuranceDTO>(`${this.apiUrl}/updateLife/${id}`, dto, {
      headers: this.authService.getAuthHeaders()
    });
  }

  getAll(): Observable<LifeInsuranceDTO[]> {
    return this.http.get<LifeInsuranceDTO[]>(`${this.apiUrl}/get`, {
      headers: this.authService.getAuthHeaders()
    });
  }

  deleteLife(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/delete/${id}`, {
      headers: this.authService.getAuthHeaders()
    });
  }

  calculatePremium(dto: LifeInsuranceDTO): Observable<number> {
    return this.http.post<number>(`${this.apiUrl}/calculate-premium`, dto, {
      headers: this.authService.getAuthHeaders()
    });
  }
}
